drivers/gpio/gpio_nrfx.c

```diff
@@ -299,45 +304,42 @@ static const struct gpio_driver_api gpio_nrfx_drv_api_funcs = {
 	.manage_callback = gpio_nrfx_manage_callback,
 };
 
-static inline uint32_t get_level_pins(const struct device *port)
-{
-	struct gpio_nrfx_data *data = get_port_data(port);
-
-	/* Take into consideration only pins that were configured to
-	 * trigger interrupts.
-	 */
-	uint32_t out = data->pin_int_en;
-
-	/* Exclude pins that trigger interrupts by edge. */
-	out &= ~data->trig_edge & ~data->double_edge;
-
-	/* The sequence above assumes that the sense field will be
-	 * configured only for these pins.  If anybody's modifying
-	 * PIN_CNF directly it won't work.
-	 */
-	return out;
-}
-
-static void cfg_level_pins(const struct device *port)
+static void cfg_level_pins(const struct device *port, uint32_t sense_levels)
 {
+#if IS_ENABLED(CONFIG_GPIO_NRF_INT_EDGE_USING_SENSE)
+#define check_pins	int_pins
+#else
+#define check_pins	level_pins
+#endif
 	const struct gpio_nrfx_data *data = get_port_data(port);
 	const struct gpio_nrfx_cfg *cfg = get_port_cfg(port);
 	uint32_t pin = 0U;
 	uint32_t bit = 1U << pin;
-	uint32_t level_pins = get_level_pins(port);
+	uint32_t int_pins = data->pin_int_en;
+	uint32_t level_pins = int_pins & (~data->trig_edge &
+					  ~data->double_edge);
 
 	/* Configure sense detection on all pins that use it. */
-	while (level_pins) {
-		if (level_pins & bit) {
+	while (check_pins) {
+		if (check_pins & bit) {
 			uint32_t abs_pin = NRF_GPIO_PIN_MAP(cfg->port_num, pin);
-			uint32_t sense = sense_for_pin(data, pin);
+			uint32_t sense;
```

Comment:
This algorithm isn't robust: you can end up configuring the pin to detect a change to the state it's already in, if the pin state changed after the interrupt was triggered, or if it changes before the sense is reconfigured.  This means notifications can be lost.

[This code](https://github.com/pabigot/nrfcxx/blob/c7088a998cde7c7fec05523d36ebed1014cc05f4/src/gpio.cc#L12) shows some of the complexities involved; see [this](https://github.com/pabigot/nrfcxx/blob/c7088a998cde7c7fec05523d36ebed1014cc05f4/include/nrfcxx/gpio.hpp#L743) for its documentation.

Reply:
>This algorithm isn't robust: you can end up configuring the pin to detect a change to the state it's already in, if the pin state changed after the interrupt was triggered

Do you mean after interrupt was triggered, but before interrupt handler was called? I don't think this is a problem in majority of use cases. Edge detection with GPIOTE also isn't able to detect ALL edges before running callback. The most important thing to do is to run a callback after SOME edge was detected.

>or if it changes before the sense is reconfigured.

That indeed is a problem, especially when interrupt handler clears the interrupt, but it is retriggered moment after that, but before reconfiguring sense. I think we can fix that if we reenable sense for all the "edge" pins just before firing callbacks.

Reply:
I've replaced reading port IN with reading port LATCH. This allows to detect edges which happen just after gpio callback was executed, but before SENSE was reconfigured.